stores/sql/metrics.go

```diff
@@ -3,79 +3,221 @@ package sql
 
 	"go.sia.tech/core/types"
 	"go.sia.tech/renterd/api"
 	"go.sia.tech/renterd/internal/sql"
 )
 
-func ContractPruneMetrics(ctx context.Context, tx sql.Tx, start time.Time, n uint64, interval time.Duration, opts api.ContractPruneMetricsQueryOpts) (metrics []api.ContractPruneMetric, _ error) {
-	rows, err := queryPeriods(ctx, tx, start, n, interval, opts)
-	if err != nil {
-		return nil, fmt.Errorf("failed to fetch contract prune metrics: %w", err)
+const (
+	contractMetricGranularity = 5 * time.Minute
+)
+
+type (
+	ContractMetricsQueryOpts struct {
+		api.ContractMetricsQueryOpts
+		IndexHint string
 	}
-	defer rows.Close()
+)
 
-	for rows.Next() {
-		var cpm api.ContractPruneMetric
+func ContractMetrics(ctx context.Context, tx sql.Tx, start time.Time, n uint64, interval time.Duration, opts ContractMetricsQueryOpts) (metrics []api.ContractMetric, err error) {
+	// define a helper function to scan a contract metric from a row.
+	scanContractMetric := func(rows *sql.LoggedRows, aggregate bool) (cm api.ContractMetric, err error) {
 		var placeHolder int64
 		var placeHolderTime time.Time
+		var placeHolderHK PublicKey
+		var placeHolderFCID FileContractID
+		var placeHolderRevisionNumber Unsigned64
+
 		var timestamp UnixTimeMS
-		if err := rows.Scan(&placeHolder, &placeHolderTime, &timestamp, (*FileContractID)(&cpm.ContractID), (*PublicKey)(&cpm.HostKey), &cpm.HostVersion, (*Unsigned64)(&cpm.Pruned), (*Unsigned64)(&cpm.Remaining), &cpm.Duration); err != nil {
-			return nil, fmt.Errorf("failed to scan contract prune metric: %w", err)
+		err = rows.Scan(
+			&placeHolder,
+			&placeHolderTime,
+			&timestamp,
+			&placeHolderFCID,
+			&placeHolderHK,
+			(*Unsigned64)(&cm.RemainingCollateral.Lo), (*Unsigned64)(&cm.RemainingCollateral.Hi),
+			(*Unsigned64)(&cm.RemainingFunds.Lo), (*Unsigned64)(&cm.RemainingFunds.Hi),
+			&placeHolderRevisionNumber,
+			(*Unsigned64)(&cm.UploadSpending.Lo), (*Unsigned64)(&cm.UploadSpending.Hi),
+			(*Unsigned64)(&cm.DownloadSpending.Lo), (*Unsigned64)(&cm.DownloadSpending.Hi),
+			(*Unsigned64)(&cm.FundAccountSpending.Lo), (*Unsigned64)(&cm.FundAccountSpending.Hi),
+			(*Unsigned64)(&cm.DeleteSpending.Lo), (*Unsigned64)(&cm.DeleteSpending.Hi),
+			(*Unsigned64)(&cm.ListSpending.Lo), (*Unsigned64)(&cm.ListSpending.Hi),
+		)
+		if err != nil {
 		}
-		cpm.Timestamp = api.TimeRFC3339(normaliseTimestamp(start, interval, timestamp))
-		metrics = append(metrics, cpm)
+
+		cm.Timestamp = api.TimeRFC3339(normaliseTimestamp(start, interval, timestamp))
+		if !aggregate {
+			cm.ContractID = types.FileContractID(placeHolderFCID)
+			cm.HostKey = types.PublicKey(placeHolderHK)
+			cm.RevisionNumber = uint64(placeHolderRevisionNumber)
+		}
+		return
 	}
 
-	return metrics, nil
+	// if a host filter is set, query periods
+	if opts.ContractID != (types.FileContractID{}) || opts.HostKey != (types.PublicKey{}) {
+		err = queryPeriods(ctx, tx, start, n, interval, opts.ContractMetricsQueryOpts, func(rows *sql.LoggedRows) error {
```

Comment:
If you are feeling adventurous you can turn this into 
```go
		metrics, err = queryPeriods(ctx, tx, start, n, interval, opts.ContractMetricsQueryOpts, func(rows *sql.LoggedRows) (api.ContractMetric, error) {
			return scanContractMetric(rows, false)
		})
```

by changing `queryPeriods` to 
```go
func queryPeriods[T any](ctx context.Context, tx sql.Tx, start time.Time, n uint64, interval time.Duration, opts interface{}, scanRowFn func(*sql.LoggedRows) (T, error)) ([]T, error) {
	if n > api.MetricMaxIntervals {
		return nil, api.ErrMaxIntervalsExceeded
	}
	params := []interface{}{
		UnixTimeMS(start),
		interval.Milliseconds(),
		UnixTimeMS(start.Add(time.Duration(n) * interval)),
		interval.Milliseconds(),
		interval.Milliseconds(),
	}

	where, err := whereClauseFromQueryOpts(opts)
	if err != nil {
		return nil, fmt.Errorf("failed to build where clause: %w", err)
	} else if len(where.params) > 0 {
		params = append(params, where.params...)
	}

	rows, err := tx.Query(ctx, fmt.Sprintf(`
		WITH RECURSIVE periods AS (
			SELECT ? AS period_start
			UNION ALL
			SELECT period_start + ?
			FROM periods
			WHERE period_start < ? - ?
		)
		SELECT %s.* FROM %s
		INNER JOIN (
		SELECT
			p.period_start as Period,
			MIN(obj.id) AS id
		FROM
			periods p
		INNER JOIN
			%s obj ON obj.timestamp >= p.period_start AND obj.timestamp < p.period_start + ?
		WHERE %s
		GROUP BY
			p.period_start
		) i ON %s.id = i.id ORDER BY Period ASC
	`, where.table, where.table, where.table, where.query, where.table), params...)
	if err != nil {
		return nil, fmt.Errorf("failed to query periods: %w", err)
	}
	defer rows.Close()

	var result []T
	for rows.Next() {
		ele, err := scanRowFn(rows)
		if err != nil {
			return nil, fmt.Errorf("failed to scan row: %w", err)
		}
		result = append(result, ele)
	}
	return result, nil
}
```